[JSC] Improve error message for calling constructor without `new`

sosukesuzuki opened the pull request:
#### dabbab2ba61ec615fba37cb099de201ff4be572c
<pre>
[JSC] Improve error message for calling constructor without `new`
<a href="https://bugs.webkit.org/show_bug.cgi?id=298927">https://bugs.webkit.org/show_bug.cgi?id=298927</a>

Reviewed by Yusuke Suzuki.

Calling constructors without `new` is invalid:

        class Foo {}
        Foo();

Current JSC throws a TypeError with the message &quot;Cannot call a class
constructor without |new|&quot; for such code.

This patch changes this error message to include the constructor name
for debugging experience.

Test: JSTests/stress/constructor-call-without-new.js

Test: JSTests/stress/constructor-call-without-new.js
* JSTests/stress/calling-non-callable-constructors.js:
(shouldThrow):
* JSTests/stress/constructor-call-without-new.js: Added.
(shouldThrow):
(throw.new.Error):
* JSTests/stress/constructor-kind-naked-should-not-be-applied-to-inner-functions.js:
(shouldThrow.Promise):
* JSTests/stress/promise-cannot-be-called.js:
(shouldThrow):
* LayoutTests/js/Promise-types-expected.txt:
* LayoutTests/js/class-syntax-call-expected.txt:
* LayoutTests/js/class-syntax-default-constructor-expected.txt:
* LayoutTests/js/script-tests/Promise-types.js:
* LayoutTests/js/script-tests/class-syntax-call.js:
* LayoutTests/js/script-tests/class-syntax-default-constructor.js:
* LayoutTests/performance-api/performance-observer-exception-expected.txt:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):

Canonical link: <a href="https://commits.webkit.org/301023@main">https://commits.webkit.org/301023@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22a9660a15cdaedf26e2f98968348f362c2666aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124708 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44380 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35115 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131550 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76630 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45084 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52950 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94880 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62935 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127662 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111532 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75449 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75029 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116813 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105714 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29918 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134216 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123228 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51556 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103355 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51968 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107753 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103127 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26254 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48502 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26775 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48544 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51430 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57227 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/156305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50823 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/156305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54179 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52518 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -533,14 +533,22 @@ BytecodeGenerator::BytecodeGenerator(VM& vm, FunctionNode* functionNode, Unlinke
+            if (!constructorName || constructorName.isEmpty())
+                emitThrowTypeError("Cannot call a constructor without |new|"_s);
+            else
+                emitThrowTypeError(Identifier::fromString(m_vm, makeString("Cannot call a constructor "_s, constructorName, " without |new|"_s)));
```
I think we should use `tryMakeString` etc. to avoid Crash for too long string.

sosukesuzuki replied:
@Constellation Thanks, but if an OOM occurs, should we throw an error, or fall back to an error message without using `makeString`?

sosukesuzuki replied:
Updated

Constellation commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -533,14 +533,22 @@ BytecodeGenerator::BytecodeGenerator(VM& vm, FunctionNode* functionNode, Unlinke
+            if (!constructorName || constructorName.isEmpty())
+                emitThrowTypeError("Cannot call a class constructor without |new|"_s);
+            else
+                emitThrowTypeError(Identifier::fromString(m_vm, makeString("Cannot call a class constructor "_s, constructorName, " without |new|"_s)));
```
Ditto.

sosukesuzuki committed:
[JSC] Improve error message for calling constructor without `new`
https://bugs.webkit.org/show_bug.cgi?id=298927

Reviewed by Yusuke Suzuki.

Calling constructors without `new` is invalid:

	class Foo {}
	Foo();

Current JSC throws a TypeError with the message "Cannot call a class
constructor without |new|" for such code.

This patch changes this error message to include the constructor name
for debugging experience.

Test: JSTests/stress/constructor-call-without-new.js

Test: JSTests/stress/constructor-call-without-new.js
* JSTests/stress/calling-non-callable-constructors.js:
(shouldThrow):
* JSTests/stress/constructor-call-without-new.js: Added.
(shouldThrow):
(throw.new.Error):
* JSTests/stress/constructor-kind-naked-should-not-be-applied-to-inner-functions.js:
(shouldThrow.Promise):
* JSTests/stress/promise-cannot-be-called.js:
(shouldThrow):
* LayoutTests/js/Promise-types-expected.txt:
* LayoutTests/js/class-syntax-call-expected.txt:
* LayoutTests/js/class-syntax-default-constructor-expected.txt:
* LayoutTests/js/script-tests/Promise-types.js:
* LayoutTests/js/script-tests/class-syntax-call.js:
* LayoutTests/js/script-tests/class-syntax-default-constructor.js:
* LayoutTests/performance-api/performance-observer-exception-expected.txt:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):

Canonical link: https://commits.webkit.org/301023@main
